android/java/org/chromium/chrome/browser/search_engines/BraveTemplateUrlServiceFactory.java

```diff
@@ -0,0 +1,97 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.chromium.chrome.browser.search_engines;
+
+import androidx.annotation.Nullable;
+import androidx.annotation.VisibleForTesting;
+
+import org.chromium.base.Log;
+import org.chromium.base.ThreadUtils;
+import org.chromium.base.annotations.NativeMethods;
+import org.chromium.chrome.browser.profiles.Profile;
+import org.chromium.components.search_engines.TemplateUrlService;
+
+import  java.lang.reflect.InvocationTargetException;
+
+public class BraveTemplateUrlServiceFactory {
+    private static final String TAG = "BraveTemplateUrlServiceFactory";
+
+    private static TemplateUrlService sNormalTemplateUrlService;
+    private static @Nullable Profile sProfile;
+
+    private BraveTemplateUrlServiceFactory() {}
+
+    /**
+     * @return The singleton instance of {@link TemplateUrlService} for profile, based on current active tab.
+     * All TemplateUrlServiceFactory.get() calls are redirected here.
+     */
+    public static TemplateUrlService get() {
+        ThreadUtils.assertOnUiThread();
+
+        if (sProfile != null) return get(sProfile);
+
+        Profile profile = null;
+
+        // Cannot import BraveActivity here due to circular dependency, need reflection all the way
```

Comment:
This indicates that we are misusing it.

Aside from this, in general I don't really get it, previously we had quite simple logic to apply current search engine depending on whether we are currently on private or regular tab. What has changed? How closing all private tabs fixing it? Frankly I would rather require user to relaunch browser to apply the setting instead of these substantial changes. As it's really hard to predict all unintended issues we may get going forward.

Reply:
Please check DM for a brief explanation of the issue. Also in addition:

> What has changed?

There are now two services instead of one, one tied to regular profile and one tied to OTR profile. Also the service tied to OTR profile will be destroyed time to time (e.g. when user closes all private tabs and open one again later). And these all happens in native and share codes with desktop so it is not possible to override.

> How closing all private tabs fixing it?

The service tied to OTR profile got terminated, and the next time a new private tab is opened, a new service is started, copying prefs from regular profile's service.

> require user to relaunch browser

That is rather a product/design decision I cannot make. But as far as the current implementation goes, I see no other way to make things work because of the desktop/native changes, so this is basically the best we can do.

Reply:
```
Profile* GetOriginalProfile() {
  return ProfileManager::GetActiveUserProfile()->GetOriginalProfile();
}
}  // namespace
static TemplateURLService* GetTemplateUrlService() {
  return TemplateURLServiceFactory::GetForProfile(GetOriginalProfile());
}
```
won't that give us a current active tab profile?

Reply:
Is that the code in `template_url_service_factory_android.cc`? That function is never called. The service is acquired with `static TemplateURLService* GetTemplateUrlService(Profile* profile)` in `brave_template_url_service_factory_android.cc` instead.

Reply:
Also it doesn't give us the service linked to active tab. `GetOriginalProfile` only return regular profile. If I remember correctly `GetActiveUserOrOffTheRecordProfile()` dpesn't work the same way on Desktop and Android, that was why Brian said we shall use Profile from Java whenever possible.

Reply:
Maybe there is a way to get it done without the layering violation? I also don't like that part much. I just saw that there is an access to profiles from native, is there a way to get current active tab from native?

Reply:
Updated with a way to iterate through all tab models and find the active one.